element/index.js

```diff
@@ -163,6 +163,27 @@ export function getWrapperDisplayName( BaseComponent, wrapperName ) {
 	return `${ upperFirst( camelCase( wrapperName ) ) }(${ displayName })`;
 }
 
+/**
+ * Given a function mapping a component to an enhanced component and modifier
+ * name, returns the enhanced component augmented with a generated displayName.
+ *
+ * @param {Function} mapComponentToEnhancedComponent Function mapping component
+ *                                                   to enhanced component.
+ * @param {string}   modifierName                    Seed name from which to
+ *                                                   generated display name.
+ *
+ * @return {WPComponent} Component class with generated display name assigned.
+ */
+export function createHigherOrderComponent( mapComponentToEnhancedComponent, modifierName ) {
+	return ( OriginalComponent ) => {
+		const EnhancedComponent = mapComponentToEnhancedComponent( OriginalComponent );
+		const { displayName = OriginalComponent.name || 'Component' } = OriginalComponent;
+		EnhancedComponent.displayName = `${ modifierName }(${ displayName })`;
```

Comment:
We had `upperFirst` & `camelCase` on top of the modifier name to make sure it always looks like the name of the component. We may enforce it at the review time. Your call.